components/mediators/cache/org.wso2.carbon.mediator.cache/src/main/java/org/wso2/carbon/mediator/cache/CacheMediator.java

```diff
@@ -354,38 +371,66 @@ private boolean processRequestMessage(MessageContext synCtx,
 				synCtx.setResponse(true);
 				opCtx.setProperty(CachingConstants.CACHED_OBJECT, cachedResponse);
 
-				SOAPEnvelope omSOAPEnv = null;
+				OMElement response = null;
 
 				try {
 					if (msgCtx.isDoingREST()) {
+
+						String replacementValue = new String(cachedResponse.getResponseEnvelope());
+						try {
+							response = AXIOMUtil.stringToOM(replacementValue);
+						} catch (XMLStreamException e) {
+							handleException("Error creating response OM from cache : " + cacheKey, synCtx);
+						}
 						if ((headerProperties = cachedResponse.getHeaderProperties()) != null) {
 
-							omSOAPEnv = SOAPMessageHelper.buildSOAPEnvelopeFromBytes(responseEnvelop, cachedResponse.isSOAP11());
-							msgCtx.removeProperty("NO_ENTITY_BODY");
+							msgCtx.removeProperty(NO_ENTITY_BODY);
 							msgCtx.removeProperty(Constants.Configuration.CONTENT_TYPE);
-							msgCtx.setProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS, headerProperties);
+							msgCtx.setProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS,
+									headerProperties);
 							msgCtx.setProperty(Constants.Configuration.MESSAGE_TYPE,
-							                   headerProperties.get(Constants.Configuration.MESSAGE_TYPE));
+									headerProperties.get(Constants.Configuration.MESSAGE_TYPE));
 						}
 
 					} else {
-						omSOAPEnv = SOAPMessageHelper.buildSOAPEnvelopeFromBytes(responseEnvelop, msgCtx.isSOAP11());
-						//finally set soap envelope by obtaining built response envelope
-						cachedResponse.setResponseEnvelope(omSOAPEnv.toString().getBytes());
+						String replacementValue = new String(cachedResponse.getResponseEnvelope());
+						try {
+							response = AXIOMUtil.stringToOM(replacementValue);
+						} catch (XMLStreamException e) {
+							handleException("Error creating response OM from cache : " + cacheKey, synCtx);
+						}
 					}
 
-					if (omSOAPEnv != null) {
-						synCtx.setEnvelope(omSOAPEnv);
+					if (response != null) {
+						// Set the headers of the message
+						if (response.getFirstElement().getLocalName().contains(HEADER)) {
```

Comment:
We have some other parts outside the second if statement, but within first statement. So maybe better to keep it as it is? 

if (msgCtx.getEnvelope().getBody().getFirstElement() != null) {
	msgCtx.getEnvelope().getBody().getFirstElement().detach();
}						msgCtx.getEnvelope().getBody().addChild(response.getFirstElement().getFirstElement());